src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs

```diff
@@ -450,7 +450,19 @@ private static bool ServerHasWebSockets(IFeatureCollection features)
             }
 
             var pipeWriterStream = new PipeWriterStream(connection.Application.Output);
-            await context.Request.Body.CopyToAsync(pipeWriterStream);
+
+            // REVIEW: Consider spliting the connection lock into a read lock and a write lock
+            // Need to think about HttpConnection.DisposeAsync and whether one or both locks would be needed
```

Comment:
You mean `HttpConnectionContext.DisposeAsync` right? I'm not sure that will do anything as the code is currently structured, so we'll have to think about it a bit.

If the context is disposed, the `PipeWriter` we're writing to [will be completed](https://github.com/aspnet/SignalR/blob/b0c4e9d0f71766340a6285ab0d5c17e01542f6be/src/Microsoft.AspNetCore.Http.Connections/HttpConnectionContext.cs#L154) and the `CopyToAsync` will start throwing exceptions. Even if we lock, that will still happen. We'd need to check some kind of "is disposed" state to prevent that. We'll need to check the `HttpConnectionContext.Status` property, which requires that current lock be taken.

Maybe it's this method that will need to take both locks? That seems unpleasant though.